query-parser/src/parse.rs

```diff
@@ -10,156 +10,191 @@
-        // operate independently.
-        choice::<[&mut Parser<Input = I, Output = FindSpec>; 4],
-                 _>([&mut try(FindSp::find_scalar()),
-                     &mut try(FindSp::find_coll()),
-                     &mut try(FindSp::find_tuple()),
-                     &mut try(FindSp::find_rel())])
-            .parse_stream(input)
-    }
+pub struct Query<I>(::std::marker::PhantomData<fn(I) -> I>);
 
+impl<I> Query<I>
+    where I: Stream<Item = edn::Value>
+{
     fn to_parsed_value<T>(r: ParseResult<T, I>) -> Option<T> {
         r.ok().map(|x| x.0)
     }
 }
 
+def_value_satisfy_parser_fn!(Query, variable, Variable, Variable::from_value);
+def_value_satisfy_parser_fn!(Query, source_var, SrcVar, SrcVar::from_value);
+
+pub struct Where<I>(::std::marker::PhantomData<fn(I) -> I>);
+
+def_value_satisfy_parser_fn!(Where, pattern_value_place, PatternValuePlace, PatternValuePlace::from_value);
+def_value_satisfy_parser_fn!(Where, pattern_non_value_place, PatternNonValuePlace, PatternNonValuePlace::from_value);
+
+def_value_parser_fn!(Where, pattern, Pattern, input, {
+    satisfy_map(|x: edn::Value| {
+        if let edn::Value::Vector(y) = x {
+            // While *technically* Datomic allows you to have a query like:
+            // [:find … :where [[?x]]]
+            // We don't -- we require at list e, a.
+            let mut p =
+                (optional(Query::source_var()),                 // src
+                 Where::pattern_non_value_place(),              // e
+                 Where::pattern_non_value_place(),              // a
+                 optional(Where::pattern_value_place()),        // v
+                 optional(Where::pattern_non_value_place()),    // tx
+                 eof())
+                .map(|(src, e, a, v, tx, _)| {
+                    let v = v.unwrap_or(PatternValuePlace::Placeholder);
+                    let tx = tx.unwrap_or(PatternNonValuePlace::Placeholder);
+
+                    // Pattern::new takes care of reversal.
+                    Pattern::new(src, e, a, v, tx)
+                });
+
+            // This is a bit messy: the inner conversion to a Pattern can
+            // fail if the input is something like
+            //
+            // ```edn
+            // [?x :foo/_reversed 23.4]
+            // ```
+            //
+            // because
+            //
+            // ```edn
+            // [23.4 :foo/reversed ?x]
+            // ```
+            //
+            // is nonsense. That leaves us with nested optionals; we unwrap them here.
+            let r: ParseResult<Option<Pattern>, _> = p.parse_lazy(&y[..]).into();
+            let v: Option<Option<Pattern>> = r.ok().map(|x| x.0);
+            v.unwrap_or(None)
+        } else {
+            None
+        }
+    }).parse_stream(input)
+});
+
+def_value_parser_fn!(Where, clauses, Vec<WhereClause>, input, {
+    // Right now we only support patterns.
```

Comment:
`:in` is elsewhere — this is about parsing `or`, `not`, predicate functions, and all the rest. But yes, I'll file the issues and leave a TODO.